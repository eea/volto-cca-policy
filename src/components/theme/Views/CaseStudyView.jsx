import React, { Fragment } from 'react';
import {
  HTMLField,
  ContentMetadata,
  LinksList,
  PublishedModifiedInfo,
  DocumentsList,
  ShareInfo,
  BannerTitle,
} from '@eeacms/volto-cca-policy/helpers';
import { Divider, Segment, Grid, Icon, Image } from 'semantic-ui-react';
import { ImageGallery } from '@eeacms/volto-cca-policy/components';

const PrimaryPhoto = (props) => {
  const { content } = props;
  const { primary_photo, primary_photo_copyright, title } = content;

  return primary_photo !== null ? (
    <div className="case-studies-review-image-wrapper">
      <Image src={primary_photo?.scales?.mini?.download} alt={title} />
      <p>{primary_photo_copyright}</p>
    </div>
  ) : null;
};

const dataDisplay = [
  {
    type: 'HTMLField',
    field: 'challenges',
    section: 'challenges_anchor',
    title: 'Challenges',
    group: 1,
  },
  {
    type: 'HTMLField',
    field: 'objectives',
    section: 'objectives_anchor',
    title: 'Objectives',
    group: 1,
  },
  {
    type: 'AdaptationOptionsItems',
    field: 'adaptationoptions',
    section: 'adapt_options_anchor',
    title: 'Adaptation Options Implemented In This Case',
    group: 1,
  },
  {
    type: 'HTMLField',
    field: 'solutions',
    section: 'solutions_anchor',
    title: 'Solutions',
    group: 1,
  },
  {
    type: 'RelevanceItems',
    field: 'relevance',
    section: 'relevance_anchor',
    title: 'Importance and Relevance of Adaptation',
    contentTitle: 'Relevance', // override the title in content section
    group: 1,
  },
  {
    type: 'HTMLField',
    field: 'stakeholder_participation',
    section: 'stake_holder_anchor',
    title: 'Stakeholder participation',
    group: 2,
  },
  {
    type: 'HTMLField',
    field: 'success_limitations',
    section: 'success_limitations_anchor',
    title: 'Success and Limiting Factors',
    group: 2,
  },
  {
    type: 'HTMLField',
    field: 'cost_benefit',
    section: 'cost_benefit_anchor',
    title: 'Costs and Benefits',
    group: 2,
  },
  {
    type: 'HTMLField',
    field: 'legal_aspects',
    section: 'legal_aspects',
    title: 'Legal Aspects',
    group: 2,
  },
  {
    type: 'HTMLField',
    field: 'implementation_time',
    section: 'implementation_time_anchor',
    title: 'Implementation Time',
    group: 2,
  },
  {
    type: 'HTMLField',
    field: 'lifetime',
    section: 'life_time_anchor',
    title: 'Life Time',
    group: 2,
  },
  {
    type: 'HTMLField',
    field: 'contact',
    section: 'contact',
    title: 'Contact',
    group: 3,
  },
  {
    type: 'LinksList',
    field: 'websites',
    section: 'websites',
    title: 'Websites',
    group: 3,
  },
  {
    type: 'HTMLField',
    field: 'source',
    section: 'source',
    title: 'Source',
    contentTitle: 'Reference', // override the title in content section
    group: 3,
  },
];

const groups = {
  1: 'Case Study Description',
  2: 'Additional Details',
  3: 'Reference Information',
};

const findSection = (title) => {
  const found = dataDisplay.filter((item) => item.title === title);
  if (found.length > 0) {
    return found[0];
  }
  return null;
};

const sectionID = (title) => {
  const found = findSection(title);
  if (found === null) {
    return title;
  }
  return found.section;
};

const PhotoGallery = (props) => {
  const { content } = props;
  const { cca_gallery } = content;

  return (
    <>
      {cca_gallery && cca_gallery.length > 0 && (
        <div className="casetstudy-gallery">
          <div className="gallery-title">
            <span>Case Study illustrations</span>
            <span> ({cca_gallery.length}) </span>
            <Icon name="ri-image-fill" />
          </div>
          <ImageGallery items={cca_gallery} />
        </div>
      )}
    </>
  );
};

const SectionsMenu = (props) => {
  const { sections, title } = props;

  return (
    <>
      {sections.length > 0 && (
        <div>
          <h4>{title}</h4>
          <ul>
            {sections.map((data, index) => (
              <li key={index}>
                <a href={'#' + sectionID(data.title)}>{data.title}</a>
              </li>
            ))}
          </ul>
        </div>
      )}
    </>
  );
};

const SectionContent = (props) => {
  const { sectionData, content } = props;

  const adaptationOptionsLinks = () => {
    let list = [];
    for (let adaptOpt of content[sectionData.field]) {
      list.push([adaptOpt['@id'], adaptOpt.title]);
    }
    return list;
  };

  const section = content[sectionData.field];

  for (var key in section) {
    if (section[key] === '') {
      section[key] = '<p>-</p>';
    }
  }

  return (
    <div id={sectionID(sectionData.title)} className="section">
      <h5 className="section-title">
        {sectionData.contentTitle !== undefined
          ? sectionData.contentTitle
          : sectionData.title}
      </h5>
      {sectionData.type === 'LinksList' ? (
        <LinksList value={content[sectionData.field]} />
      ) : sectionData.type === 'AdaptationOptionsItems' ? (
        <LinksList
          value={adaptationOptionsLinks()}
          withText={true}
          isInternal={true}
        />
      ) : sectionData.type === 'RelevanceItems' ? (
        content[sectionData.field].map((relevanceItem, index) => (
          <p key={index}>{relevanceItem.title}</p>
        ))
      ) : (
        <HTMLField value={content[sectionData.field]} />
      )}
    </div>
  );
};

function CaseStudyView(props) {
  const { content } = props;
  const { cca_files, long_description } = content;
  const hasFiles = cca_files && cca_files.length > 0;

  const hasValue = (field) => {
    if (!content.hasOwnProperty(field)) {
      return false;
    }
    if (content[field] === undefined || content[field] === null) {
      return false;
    }
    if (Array.isArray(content[field]) && content[field].length === 0) {
      return false;
    }
    return true;
  };

  const usedSections = (group) => {
    return dataDisplay.filter(
      (data) => data.group === group && hasValue(data.field),
    );
  };

  return (
    <div className="db-item-view case-study-view">
      <BannerTitle content={{ ...content, image: '' }} type="Case Studies" />

      <div className="ui container">
        <Grid columns="12">
          <div className="row">
            <Grid.Column
              mobile={12}
              tablet={12}
              computer={8}
              className="col-left"
            >
              <PrimaryPhoto {...props} />
              <HTMLField value={long_description} />

              <Divider />
              <div className="adaptation-details">
                <Grid columns="12">
                  <Grid.Column mobile={12} tablet={12} computer={4}>
                    <SectionsMenu
                      sections={usedSections(1)}
                      title={groups['1']}
                    />
                  </Grid.Column>
                  <Grid.Column mobile={12} tablet={12} computer={4}>
                    <SectionsMenu
                      sections={usedSections(2)}
                      title={groups['2']}
                    />
                  </Grid.Column>
                  <Grid.Column mobile={12} tablet={12} computer={4}>
                    <SectionsMenu
                      sections={usedSections(3)}
                      title={groups['3']}
                    />
                  </Grid.Column>
                </Grid>
              </div>

              <Divider />

              {[1, 2, 3].map(
                (groupID, index) =>
                  usedSections(groupID).length > 0 && (
                    <Fragment key={index}>
                      <h2>{groups[groupID]}</h2>
                      {usedSections(groupID).map((data, index) => (
                        <SectionContent
                          sectionData={data}
                          content={content}
                          key={index}
                        />
                      ))}
                      {groupID !== 3 ? <Divider /> : null}
                    </Fragment>
                  ),
              )}

              <PublishedModifiedInfo {...props} />
              <Divider />

              <p>
                Please contact us for any other enquiry on this Case Study or to
                share a new Case Study (email{' '}
                <span className="link-mailto">
                  <a
                    href="mailto: climate.adapt@eea.europa.eu"
                    target="_blank"
                    rel="noreferrer"
                  >
                    climate.adapt@eea.europa.eu
                  </a>
                </span>
                )
              </p>
              <ShareInfo {...props} />
            </Grid.Column>
            <Grid.Column
              mobile={12}
              tablet={12}
              computer={4}
              className="col-right"
            >
              <PhotoGallery {...props} />
              <Segment>
                <ContentMetadata {...props} />
              </Segment>
              {hasFiles && (
                <Segment>
                  <DocumentsList {...props} />
                </Segment>
              )}
            </Grid.Column>
          </div>
        </Grid>
      </div>
    </div>
  );
}

export default CaseStudyView;
