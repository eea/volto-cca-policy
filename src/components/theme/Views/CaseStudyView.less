.db-item-view.case-study-view {
  .has-logo {
    justify-content: unset;

    img.db-logo {
      margin-left: auto;
    }
  }
}
