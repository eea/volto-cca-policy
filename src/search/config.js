import { mergeConfig } from '@eeacms/search';

import facets from './facets';

// import views from './views';
// import filters from './filters';
// import vocabs from './vocabulary';

const getClientProxyAddress = () => {
  const url = new URL(window.location);
  url.pathname = '';
  url.search = '';
  return url.toString();
};

const ccaConfig = {
  title: 'ClimateAdapt Main',
};

export default function installMainSearch(config) {
  const envConfig = process.env.RAZZLE_ENV_CONFIG
    ? JSON.parse(process.env.RAZZLE_ENV_CONFIG)
    : ccaConfig;

  const pjson = require('@eeacms/volto-cca-policy/../package.json');

  envConfig.app_name = pjson.name;
  envConfig.app_version = pjson.version;

  config.searchui.ccaSearch = {
    ...mergeConfig(envConfig, config.searchui.globalsearchbase),
    elastic_index: '_es/globalsearch',
    index_name: 'data_searchui',
    host: process.env.RAZZLE_ES_PROXY_ADDR || 'http://localhost:3000',
  };

  const { ccaSearch } = config.searchui;

  ccaSearch.permanentFilters.push({
    term: {
      cluster_name: 'cca',
    },
  });
  ccaSearch.permanentFilters.push({
    term: {
      cca_include_in_mission: 'true',
    },
  });

  ccaSearch.facets = facets;

  ccaSearch.initialView.tilesLandingPageParams.sections = [
    {
      id: 'types',
      title: 'Types',
      facetField: 'objectProvides',
      sortOn: 'alpha',
      icon: {
        family: 'Content types',
      },
    },
    // {
    //   id: 'topics',
    //   title: 'Topics',
    //   facetField: 'topic',
    //   sortOn: 'alpha',
    //   whitelist: [
    //     'Agriculture and food',
    //     'Air pollution',
    //     'Bathing water quality',
    //     'Biodiversity',
    //     'Bioeconomy',
    //     'Buildings and construction',
    //     'Chemicals',
    //     'Circular economy',
    //     'Climate change adaptation',
    //     'Climate change mitigation',
    //     'Electric vehicles',
    //     'Energy',
    //     'Energy efficiency',
    //     'Environmental health impacts',
    //     'Environmental inequalities',
    //     'Extreme weather',
    //     'Fisheries and aquaculture',
    //     'Forests and forestry',
    //     'Industry',
    //     'Land use',
    //     'Nature protection and restoration',
    //     'Nature-based solutions',
    //     'Noise',
    //     'Plastics',
    //     'Pollution',
    //     'Production and consumption',
    //     'Renewable energy',
    //     'Resource use and materials',
    //     'Road transport',
    //     'Seas and coasts',
    //     'Soil',
    //     'Sustainability challenges',
    //     'Sustainability solutions',
    //     'Sustainable finance',
    //     'Textiles',
    //     'Transport and mobility',
    //     'Urban sustainability',
    //     'Waste and recycling',
    //     'Water',
    //   ],
    // },
    // {
    //   id: 'countries',
    //   title: 'Countries',
    //   facetField: 'spatial',
    //   filterType: 'any:exact',
    //   sortOn: 'alpha',
    //   maxPerSection: 100,
    //   whitelist: [
    //     'Austria',
    //     'Belgium',
    //     'Bulgaria',
    //     'Croatia',
    //     'Cyprus',
    //     'Czechia',
    //     'Denmark',
    //     'Estonia',
    //     'Finland',
    //     'France',
    //     'Germany',
    //     'Greece',
    //     'Hungary',
    //     'Iceland',
    //     'Ireland',
    //     'Italy',
    //     'Latvia',
    //     'Liechtenstein',
    //     'Lithuania',
    //     'Luxembourg',
    //     'Malta',
    //     'Netherlands',
    //     'Norway',
    //     'Poland',
    //     'Portugal',
    //     'Romania',
    //     'Slovakia',
    //     'Slovenia',
    //     'Spain',
    //     'Sweden',
    //     'Switzerland',
    //     'TÃ¼rkiye',
    //     'Albania',
    //     'Bosnia and Herzegovina',
    //     'Kosovo',
    //     'Montenegro',
    //     'North Macedonia',
    //     'Serbia',
    //   ],
    //   icon: {
    //     family: 'CountryFlags',
    //     className: 'facet-option-icon',
    //   },
    // },
    // {
    //   id: 'language',
    //   title: 'Languages',
    //   facetField: 'language',
    //   sortOn: 'custom',
    //   sortOrder: 'asc',
    // },
    // {
    //   id: 'website',
    //   title: 'Websites',
    //   facetField: 'cluster_name',
    //   sortOn: 'count',
    //   sortOrder: 'desc',
    //   icon: {
    //     family: 'Sources',
    //     className: 'facet-option-icon',
    //   },
    // },
  ];

  if (typeof window !== 'undefined') {
    config.searchui.ccaSearch.host =
      process.env.RAZZLE_ES_PROXY_ADDR || getClientProxyAddress();
  }
  // console.log(config.searchui.ccaSearch);

  return config;
}
