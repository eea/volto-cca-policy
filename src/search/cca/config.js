import { mergeConfig } from '@eeacms/search';
import { getClientProxyAddress } from '../utils';
import { vocab } from '../vocabulary';

import facets from './facets';
import views from './views';
import { download_fields } from '../../constants';

const ccaConfig = {
  title: 'ClimateAdapt Main',
  ...views,
};

export const clusters = {
  name: 'op_cluster',
  // field: 'objectProvides',
  clusters: [
    {
      name: 'Climate-ADAPT',
      icon: { name: 'file text' },
      values: ['climate'],
    },
    {
      name: 'Health Observatory',
      icon: { name: 'compass' },
      values: ['observatory'],
    },
    {
      name: 'Mission Portal',
      icon: { name: 'compass' },
      values: ['mission'],
    },
  ].map((cluster) => ({
    ...cluster,
    defaultResultView: 'ClusterHorizontalCardItem',
  })),
};

const cca_build_runtime_mappings = {};
cca_build_runtime_mappings['op_cluster'] = {
  type: 'keyword',
  script: {
    source:
      "def objectProvides = ['Adaptation option','Case study','Guidance','Video','Indicator','Information portal','Organisation','Publication reference','Research and knowledge project','Video','Tool'];def vals = doc['objectProvides'];for (defVal in objectProvides) {for (objVal in vals) {if (objVal.contains(defVal)) {emit('Climate-ADAPT')}}}emit('_all_');if (doc['cca_include_in_search_observatory'][0]) {for (defVal in objectProvides) {for (objVal in vals) {if (objVal.contains(defVal)){emit('Health Observatory')}}}}  if (doc['cca_include_in_mission.keyword'][0].toString() == 'true') {emit('Mission Portal')} else {if (doc['id.keyword'].length==1) { if(doc['id.keyword'][0].toString().contains('climate-adapt.eea.europa.eu/en/mission')) {emit('Mission Portal')}}}",
  },
};

export default function installMainSearch(config) {
  const envConfig = ccaConfig;
  const pjson = require('@eeacms/volto-cca-policy/../package.json');

  envConfig.app_name = pjson.name;
  envConfig.app_version = pjson.version;

  config.searchui.ccaSearch = {
    ...mergeConfig(envConfig, config.searchui.globalsearchbase),
    elastic_index: '_es/globalsearch',
    index_name: 'data_searchui',
    host: process.env.RAZZLE_ES_PROXY_ADDR || 'http://localhost:3000',
    runtime_mappings: cca_build_runtime_mappings,
    vocab,
  };

  config.searchui.ccaSearch.download_fields = download_fields;

  const { ccaSearch } = config.searchui;

  ccaSearch.contentSectionsParams = {
    enable: true,
    sectionFacetsField: 'op_cluster',
    sections: clusters.clusters,
    clusterMapping: Object.assign(
      {},
      ...clusters.clusters.map(({ name, values }) =>
        Object.assign({}, ...values.map((v) => ({ [v]: name }))),
      ),
    ),
  };

  ccaSearch.permanentFilters.push({
    term: {
      cluster_name: 'cca',
    },
  });

  ccaSearch.facets = facets;

  ccaSearch.initialView.tilesLandingPageParams.sections = [
    {
      id: 'types',
      title: 'Types',
      facetField: 'objectProvides',
      sortOn: 'alpha',
      icon: {
        family: 'Content types',
      },
    },
    {
      id: 'cca_geographic_countries',
      title: 'Countries',
      facetField: 'cca_geographic_countries.keyword',
      // filterType: 'any:exact',
      sortOn: 'alpha',
      // maxPerSection: 100,
      whitelist: [
        'Austria',
        'Belgium',
        'Bulgaria',
        'Croatia',
        'Cyprus',
        'Czechia',
        'Denmark',
        'Estonia',
        'Finland',
        'France',
        'Germany',
        'Greece',
        'Hungary',
        'Iceland',
        'Ireland',
        'Italy',
        'Latvia',
        'Liechtenstein',
        'Lithuania',
        'Luxembourg',
        'Malta',
        'Netherlands',
        'Norway',
        'Poland',
        'Portugal',
        'Romania',
        'Slovakia',
        'Slovenia',
        'Spain',
        'Sweden',
        'Switzerland',
        'TÃ¼rkiye',
        'Albania',
        'Bosnia and Herzegovina',
        'Kosovo',
        'Montenegro',
        'North Macedonia',
        'Serbia',
      ],
      icon: {
        family: 'CountryFlags',
        className: 'facet-option-icon',
      },
    },
  ];

  if (typeof window !== 'undefined') {
    config.searchui.ccaSearch.host =
      process.env.RAZZLE_ES_PROXY_ADDR || getClientProxyAddress();
  }
  let fieldsMatchAll = [
    'cca_climate_impacts.keyword',
    'cca_adaptation_sectors.keyword',
    'cca_adaptation_elements.keyword',
    'cca_key_type_measure.keyword',
  ];
  config.searchui.ccaSearch.facets.forEach((facet) => {
    if (fieldsMatchAll.includes(facet.field)) {
      facet.filterType = 'all';
    }
  });

  return config;
}
