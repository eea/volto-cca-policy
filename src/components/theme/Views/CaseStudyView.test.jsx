import React from 'react';
import { MemoryRouter } from 'react-router-dom';
import configureStore from 'redux-mock-store';
import '@testing-library/jest-dom/extend-expect';
import { Provider } from 'react-intl-redux';
import CaseStudyView from './CaseStudyView';
import renderer from 'react-test-renderer';
import config from '@plone/volto/registry';

config.blocks = {
  blocksConfig: {
    title: {
      view: () => <div>Title Block Component</div>,
    },
  },
};

const mockStore = configureStore();

jest.mock('semantic-ui-react', () => ({
  ...jest.requireActual('semantic-ui-react'),
}));

describe('CaseStudyView', () => {
  it('should render the component', () => {
    const content = {
      title: 'CaseStudyView',
      cca_gallery: [],
      cca_published: '2022-06-24T12:52:50+00:00',
      geochars:
        '{\r\n "geoElements":{"element":"GLOBAL",\r\n "macrotrans":null,"biotrans":null,"countries":[],\r\n "subnational":[],"city":""}}',
    };

    const store = mockStore({
      userSession: { token: '1234' },
      intl: {
        locale: 'en',
        messages: {},
      },
    });

    const component = renderer.create(
      <Provider store={store}>
        <MemoryRouter>
          <CaseStudyView content={content} />
        </MemoryRouter>
      </Provider>,
    );
    const json = component.toJSON();
    expect(json).toMatchSnapshot();
  });
});
