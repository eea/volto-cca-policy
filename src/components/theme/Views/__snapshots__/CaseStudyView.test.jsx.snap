// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CaseStudyView should render the component 1`] = `
<div
  className="db-item-view case-study-view"
>
  <div>
    Title Block Component
  </div>
  <div
    className="ui container"
  >
    <div
      className="ui twelve column grid"
    >
      <div
        className="row"
      >
        <div
          className="twelve wide computer twelve wide mobile twelve wide tablet column"
        >
          <div
            className="case-studies-review-image-wrapper"
          >
            <img
              alt="CaseStudyView"
              className="ui image"
            />
            <p />
          </div>
          <div
            dangerouslySetInnerHTML={
              Object {
                "__html": undefined,
              }
            }
          />
        </div>
        <div
          className="three wide computer twelve wide mobile twelve wide tablet column"
        >
          <div
            className="casetstudy-gallery"
          />
        </div>
      </div>
    </div>
    <div
      className="adaptation-details"
    />
    <div
      className="ui divider"
    />
    <div
      className="published-modified-info"
    >
      <p>
        <span>
          <strong>
            Published in Climate-ADAPT
          </strong>
           
          Invalid Date
        </span>
        <span>
             -   
        </span>
        <span>
          <strong>
            Last Modified in Climate-ADAPT
          </strong>
           
          Invalid Date
        </span>
      </p>
    </div>
    <div
      className="ui divider"
    />
    <p>
      Please contact us for any other enquiry on this Case Study or to share a new Case Study (email
       
      <span
        className="link-mailto"
      >
        <a
          href="mailto: climate.adapt@eea.europa.eu"
          rel="noreferrer"
          target="_blank"
        >
          climate.adapt@eea.europa.eu
        </a>
      </span>
      )
    </p>
    <div
      className="share-info"
    >
      <a
        href="/help/share-your-info"
      >
        <button
          className="ui button primary icon labeled"
        >
          Share your information
          <i
            aria-hidden="true"
            className="arrow right icon"
          />
        </button>
      </a>
    </div>
    <div
      className="content-box"
    >
      <div
        className="content-box-inner"
      >
        <div
          className="ui segment"
        >
          <div
            className="content-metadata"
          >
            <h5>
              Geographic characterisation:
            </h5>
            <div
              className="geochar"
            >
              <p>
                Global
              </p>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
